Add automatic logging of e2e flakes, and automatic restarts

nithusha21 opened the pull request:
## Overview

1. This PR fixes or fixes part of N/A
2. This PR does the following: We track all e2e flakes in a sheet. This PR edits the e2e tests runner to interact with the sheet to automatically detect if a failure is a flake. If so, it will automatically log that in the sheet, and restart the test upto 3 runs. This helps with better management and prioritization of the list of flakes, and automatically restarts, making the flakes less annoying for contributors.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

nithusha21 committed:
add automatic restarts for flakes, and automatic logging in sheets

nithusha21 committed:
use service accounts instead of API keys, and download the credentials

nithusha21 committed:
some lint fixes

nithusha21 committed:
add encryption, and fix conditions for checking for flakes

nithusha21 committed:
Merge branch 'develop' of github.com:nithusha21/oppia into custom-e2e-logging

nithusha21 committed:
add python dependencies

nithusha21 committed:
add more pip dependencies

nithusha21 committed:
fix lint issues

nithusha21 committed:
fix more lint issues

nithusha21 committed:
fix more lint issues

nithusha21 committed:
reorder codeowner files

nithusha21 committed:
fix import issues, and add missing lib to installation

nithusha21 committed:
exclude encrypted file from linting

kevintab95 commented:
Thanks @nithusha21! I don't think my review is required here -- so unassigning myself.

U8NWXD commented:
@nithusha21 since we still have to fix the backend tests, can this be made a draft?

U8NWXD committed:
Fix backend tests

nithusha21 committed:
Move installation to oppia-tools

nithusha21 committed:
Merge branch 'develop' of github.com:oppia/oppia into custom-e2e-logging

nithusha21 committed:
fix all lint issues, and import issues

nithusha21 committed:
fix lint issues

nithusha21 committed:
Add simple crpyt to path

nithusha21 committed:
add google-api-client to path

nithusha21 committed:
add missing library

nithusha21 committed:
Merge branch 'develop' of github.com:oppia/oppia into custom-e2e-logging

nithusha21 committed:
add pkg_resources

nithusha21 committed:
remove changes from linter

nithusha21 committed:
trying to debug env on CI

nithusha21 committed:
more debugging

nithusha21 committed:
more debugging

nithusha21 committed:
fix config

nithusha21 committed:
fix indent

U8NWXD committed:
Merge branch 'develop' into custom-e2e-logging

seanlip commented:
Note: I'm cancelling the Travis build on this PR temporarily because I don't think this will pass without #10835 going in. Will restart once it's in (if someone else hasn't done so by then).

seanlip commented:
Though given that backend/lint checks and a bunch of other e2e tests are failing anyway, it might be worth fixing those first before rerunning actually...

U8NWXD committed:
Fix lint errors

U8NWXD commented:
@nithusha21 It looks like the import errors aren't fully fixed yet. I'm still seeing `E: 75, 0: Unable to import 'googleapiclient.discovery' ` from the lint checks

U8NWXD committed:
Fix imports for linter

U8NWXD commented:
@nithusha21 I think I fixed the import issues: we needed to setup the google import paths in the linter too. For reference, this is the log of the failing lint check that we were seeing before: https://github.com/oppia/oppia/pull/10654/checks?check_run_id=1212591348

nithusha21 committed:
Merge branch 'develop' of github.com:oppia/oppia into custom-e2e-logging

nithusha21 committed:
change installation

nithusha21 committed:
Merge branch 'custom-e2e-logging' of github.com:nithusha21/oppia into custom-e2e-logging

nithusha21 committed:
fix linter

U8NWXD committed:
Merge branch 'develop' into custom-e2e-logging

U8NWXD commented:
The backend and lint tests are failing like this:

```
 Checking whether coverage is installed in /home/runner/work/oppia/oppia/../oppia_tools
    include_load_tests=include_load_tests)
  File "/home/runner/work/oppia/oppia/scripts/run_backend_tests.py", line 225, in _get_all_test_targets
    _get_test_target_classes(os.path.join(subroot, f)))
  File "/home/runner/work/oppia/oppia/scripts/run_backend_tests.py", line 195, in _get_test_target_classes
    python_module = importlib.import_module(test_target_path)
  File "/opt/hostedtoolcache/Python/2.7.18/x64/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "scripts/linters/other_files_linter_test.py", line 30, in <module>
    from . import pre_commit_linter
  File "scripts/linters/pre_commit_linter.py", line 136, in <module>
    'google-api-core-%s' % common.GOOGLE_API_CORE_VERSION),
AttributeError: 'module' object has no attribute 'GOOGLE_API_CORE_VERSION'
```

I'm not sure why this is happening since this works fine locally:

```python
>>> from scripts.linters import pre_commit_linter
>>> from scripts import common
>>> common.GOOGLE_API_CORE_VERSION
u'1.15.0'
```

seanlip commented:
Just a note: the same error is reproducible for me locally using the command `python -m scripts.linters.pre_commit_linter --path=. --verbose` on this branch. It does not arise on develop.

seanlip commented:
I do not see a GOOGLE_API_CORE_VERSION in the common.py on this branch. @U8NWXD, is it possible that you have unpushed changes that are staged locally but not yet pushed to GitHub?

U8NWXD committed:
Merge branch 'custom-e2e-logging' of github.com:nithusha21/oppia into custom-e2e-logging

U8NWXD commented:
@seanlip I think the issue is actually the reverse. @nithusha21 removed `GOOGLE_API_CORE_VERSION` from `scripts.common` in becbbcf90ca6dd3a9d58825c4bd9cee031417de5 earlier today. Nithesh, could you bring us up to speed on why you made that change?

DubeySandeep commented:
@nithusha21 @U8NWXD, Any update on this PR? Also, is there any rough estimate on when this PR will be merged?

U8NWXD committed:
Remove Google Import Code from Linter

becbbcf changed the way imports are handled, so we don't need the
special path-modifying code in the linter script anymore.

U8NWXD committed:
Write backend tests for run_e2e_tests functions

U8NWXD committed:
Fix lint errors

U8NWXD commented:
I added tests to cover the new functions in `run_e2e_tests.py`. We still need tests to cover the new logic in `main()`

nithusha21 committed:
Merge branch 'custom-e2e-logging' of github.com:nithusha21/oppia into custom-e2e-logging

nithusha21 committed:
Merge branch 'develop' of github.com:oppia/oppia into custom-e2e-logging

nithusha21 committed:
fix linting issues

nithusha21 committed:
Fix backend test coverage

nithusha21 committed:
fix lint issue

nithusha21 committed:
remove unnecessary change

nithusha21 committed:
fix lint problems

nithusha21 committed:
fix lint problems

nithusha21 committed:
fix import issue

U8NWXD committed:
Fix lint errors

U8NWXD committed:
Merge branch 'develop' into custom-e2e-logging

U8NWXD commented:
@nithusha21 I resolved the merge conflict in `requirements.txt`. Could you check my changes? In particular, I upgraded `google-api-core[grpc]` to 1.22.4 to match changes on develop

U8NWXD committed:
Merge branch 'develop' into custom-e2e-logging

U8NWXD commented:
I've run into an interesting problem on this. At accfe4e the e2e tests were running fine, but by de88a42, the tests were broken failing to import from `google.oauth2`. I think this is because when I fixed the lint errors, I put the `google.oauth2` imports above the `python_utils` import, and in `python_utils` we set `sys.path` ([link](https://github.com/nithusha21/oppia/blob/custom-e2e-logging/python_utils.py#L27-L29)):

```python
_FUTURE_PATH = os.path.join(
    os.getcwd(), 'third_party', 'python_libs')
sys.path.insert(0, _FUTURE_PATH)
```

I think to fix this we'll need to wait for #11018 to go in, which fixes the linter to let us put the `google.oauth2` import below the `python_utils` import. Then we can fix the import order, which should fix the tests.

seanlip commented:
If you need this to get in soonish, maybe just cherrypick or redo the changes from #11018 here?

There'll be merge conflicts down the road, but if getting this in is worth it, then that's one way to speed things up.

seanlip commented:
@Hudda @vojtechjelinek @DubeySandeep could you please review this ASAP, this is an important PR for the e2e team. Thanks!

U8NWXD committed:
Fix Import Order

We need to import `google.oauth2` after we import `python_utils` because
importing `python_utils` sets `sys.path` as needed for `google.oauth2`.
This commit replicates changes from #11018 that fix the linter to
recognize google.oauth2 as a third-party library.

U8NWXD commented:
> If you need this to get in soonish, maybe just cherrypick or redo the changes from #11018 here?
> 
> There'll be merge conflicts down the road, but if getting this in is worth it, then that's one way to speed things up.

done in 24adaff

Hudda reviewed (approved):
LGTM

DubeySandeep commented on app_dev.yaml:
```diff
@@ -300,6 +300,8 @@ skip_files:
 - third_party/python_libs/google/appengine/
 - third_party/python_libs/google/net/
+- third_party/python_libs/google/oauth2/
+- third_party/python_libs/google/api_core/
```
We are adding 4 new lib but removing 2 dirs here? Is that expected?

nithusha21 replied:
Fixed. The path was incorrect.

DubeySandeep commented on requirements.in:
```diff
@@ -27,3 +27,7 @@ requests-toolbelt==0.9.1
 google-cloud-tasks==1.5.0
 googleapis_common_protos==1.52.0
 webapp2==3.0.0b1
+google-auth==1.21.1
```
Optional: Nit: Should we add this in sorted order*  (Though it's only sorted till line 22)?

nithusha21 replied:
Done

DubeySandeep commented on requirements.txt:
```diff
@@ -30,10 +33,12 @@ graphy==1.0.0             # via -r requirements.in, googleappenginemapreduce
 grpc-google-iam-v1==0.12.3  # via google-cloud-tasks
 grpcio==1.32.0            # via google-api-core, googleapis-common-protos, grpc-google-iam-v1
 html5lib==1.0.1           # via -r requirements.in
+httplib2==0.18.1          # via google-api-python-client, google-auth-httplib2
```
Should we also remove this from the build container? (i.e, add this in -ski-file list of app.yaml)

nithusha21 replied:
Done

DubeySandeep commented on requirements.txt:
```diff
@@ -30,10 +33,12 @@ graphy==1.0.0             # via -r requirements.in, googleappenginemapreduce
 mock==3.0.5               # via googleappenginemapreduce
 mox==0.5.3                # via googleappenginemapreduce
 mutagen==1.43.0           # via -r requirements.in
+oauthlib==3.1.0           # via requests-oauthlib
```
Ditto here and elsewhere*

nithusha21 replied:
Done

DubeySandeep commented on scripts/run_e2e_tests.py:
```diff
@@ -33,6 +33,22 @@
+    os.getcwd(), 'third_party', 'python_libs')
+sys.path.insert(0, _PYTHON_LIBS_PATH)
+
+import googleapiclient.discovery # isort:skip  pylint: disable=wrong-import-position, wrong-import-order
```
We should move the google spreadsheet related code to a different file? (We can do it later if required but make sure to create an issue for the same!)

nithusha21 replied:
Created https://github.com/oppia/oppia/issues/11038 to resolve all the below comments as I don't have time right now to perform this refactor. I will do that when I am back.

DubeySandeep replied:
Thanks for creating ths issue, I'm working on it and I'll create a separate PR.

DubeySandeep commented on scripts/run_e2e_tests.py:
```diff
@@ -518,7 +534,65 @@ def cleanup_portserver(portserver_process):
+    """
+    sheet_id = os.getenv('FLAKY_E2E_TEST_SHEET_ID')
+    flaky_tests_list = []
+    if sheet_id is not None:
```
```
if sheet_id is None:
    return [] 

restult = ....
```
[To avoid lots of code inside if branch!]

DubeySandeep commented on scripts/run_e2e_tests.py:
```diff
@@ -33,6 +33,22 @@
+
+import googleapiclient.discovery # isort:skip  pylint: disable=wrong-import-position, wrong-import-order
+
+MAXIMUM_RUNS = 3
```
Maybe MAX_RETRY_COUNT or something around that?

nithusha21 replied:
Done

DubeySandeep commented on scripts/run_e2e_tests.py:
```diff
@@ -559,10 +633,80 @@ def main(args=None):
+
+    flaky_tests_list = []
+    google_auth_decode_password = os.getenv('GOOGLE_AUTH_DECODE_PASSWORD')
+    if google_auth_decode_password is not None:
```
Should we move the code 647-660 to a function get_flkay_test_data_sheet?

DubeySandeep commented on scripts/run_e2e_tests.py:
```diff
@@ -559,10 +633,80 @@ def main(args=None):
+    """Run tests, rerunning at most MAXIMUM_RUNS times if they flake."""
+    for _ in python_utils.RANGE(MAXIMUM_RUNS):
+        flake_state = run_tests(args=args)
+        if flake_state != 'flake':
```
Should we structure it this way:

```
get_flkay_test_data_sheet:
    data = get_flkay_test_data_sheet()
    return get_flaky_tests_data_from_sheets(data)

is_test_output_flaky(output):
    logs = get_flkay_test_logs()
    for log in logs:
         if is_output_similar_to_log(output, log)
             return True
    return False

main:
    return_code = 1
    Loop max_tries:
         output, return_code = run_test()
         if is_test_output_faky(output):
             update_falkey_test_count() [OR do it this inside the is_test_output_flkay() call]
             print 'The test result looks flaky, rerunning the tests (%<attempt number>)'
         else:
             break
 
    sys.exist(return_code)
```

[This is just to make the code structured and readable!]

DubeySandeep reviewed (changes requested):
I took a pass and left a few comments. Most of the comments are related to code structure, if required we can resolve that comment after getting this PR merged as this PR will help every other contributors. Let me know if you have any question/suggestion. Thanks!

vojtechjelinek reviewed (approved):
LGTM!

U8NWXD commented:
I'm seeing an import error again, but only on TravisCI:

`ImportError: No module named oauth2`

seanlip commented:
This works on my local machine (or at least, it gets past the point on Travis where it's currently stuck). Per @kevintab95 I'm going to try closing and reopening this PR in order to (hopefully) break the cache (assuming caching is the issue).

@U8NWXD, if this persists, maybe you could just test out (on a separate branch) what happens for a single Travis test, and maybe add logs as needed to get more insight into what's going on?

nithusha21 committed:
Resolve some comments

DubeySandeep committed:
Merge branch 'custom-e2e-logging' of https://github.com/nithusha21/oppia into custom-e2e-logging

DubeySandeep committed:
Updated with upstream develop

U8NWXD committed:
Merge branch 'develop' into custom-e2e-logging

seanlip commented:
Closing -- superseded by https://github.com/oppia/oppia/pull/11056.
